File: answers/lintcode/Maximum Product Subarray.cpp
class Solution {
public:
    /**
     * @param nums: a vector of integers
     * @return: an integer
     */
    int maxProduct(vector<int>& nums) {
        // write your code here
        vector<int> max_dp(nums.size(), INT_MIN);
        vector<int> min_dp(nums.size(), INT_MAX);
        int max_product = INT_MIN;
        for (size_t i = 0; i < nums.size(); ++ i) {
            max_dp[i] = max(nums[i], (i > 0 ? max(max_dp[i - 1] * nums[i], min_dp[i - 1] * nums[i]) : INT_MIN));
            min_dp[i] = min(nums[i], (i > 0 ? min(max_dp[i - 1] * nums[i], min_dp[i - 1] * nums[i]) : INT_MAX));
            max_product = max(max_product, max_dp[i]);
        }
        return max_product;
    }
};

File: answers/leetcode/Maximum Product Subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        max_total = nums[0]
        max_product = nums[0]
        min_product = nums[0]
        for i in range(1, len(nums)):
            temp = max(max_product * nums[i], min_product * nums[i], nums[i])
            min_product = min(max_product * nums[i], min_product * nums[i], nums[i])
            max_product = temp
            max_total = max(max_total, max_product)
        return max_total
